### PR TITLE
[Zeppelin-2839] permissions for folders

### PR DESCRIPTION
**Pull request by tinkoff-dwh**
### What is this PR for?
Adds permissions for folders.
Folder with permissions is way to unite notes with the same permissions and change the permissions together.

### What type of PR is it?
[Feature]

### What is the Jira issue?
https://issues.apache.org/jira/browse/ZEPPELIN-2839

### How should this be tested?
folderperm2

### Questions:
* Does the licenses files need update? No
* Is there breaking changes for older versions? No
* Does this needs documentation? Yes, added

### CODE REVIEW

**Comment by tinkoff-dwh** [comment]
Ready to review (red for other tests )

**Comment by tinkoff-dwh** [comment]
Ready to review

**Comment by 1ambda** [comment]
@tinkoff-dwh Thanks for this useful feature. I heard that some users asked about this features in SO and twitters. Let me test and comment. It takes few days.

**Comment by 1ambda** [comment]
and we have java 9 and Hadoop 3 increased the minimum java version from 7 to 8. Should we start to discuss using JDK 8 by default?

File: docs/setup/security/notebook_authorization.md [hunk 1]
```diff
@@ -81,3 +81,11 @@ For example, before executing a **Read** operation, it checks if the user and th
 Zeppelin executes a [REST API call](https://github.com/apache/zeppelin/blob/master/zeppelin-server/src/main/java/org/apache/zeppelin/rest/NotebookRestApi.java) for the notebook permission information.
 In the backend side, Zeppelin gets the user information for the connection and allows the operation if the users and groups
 associated with the current user have at least one entity that belongs to owner entities for the notebook.
+
+## Permissions for folders
+Folder with permissions is way to unite notes with the same permissions and change the permissions together.
+
+Folder without permissions (with empty permissions) continues to work like common folder.
+
+Folder with permissions sets permissions to all its content (subfolders, notes). Nobody can change permissions on content of folder with permissions. Permissions are regulated by folder.
```
Comment (1ambda) [inline docs/setup/security/notebook_authorization.md]: > Nobody can change permissions on content of folder with permissions

including admin?

File: docs/setup/security/notebook_authorization.md [hunk 2]
```diff
@@ -81,3 +81,11 @@ For example, before executing a **Read** operation, it checks if the user and th
...
+Only owners of the folder can change its permissions. If permissions is empty (folder without permissions), anybody can set permissions. Owners of the folder can remove all permissions, then folder becomes 'folder without permissions'.
```
Comment (1ambda) [inline docs/setup/security/notebook_authorization.md]: It's quite tricky to understand the total view. Could you write a table for the simplified view?

File: docs/setup/security/notebook_authorization.md [hunk 3]
```diff
@@ -81,3 +81,11 @@ For example, before executing a **Read** operation, it checks if the user and th
 Zeppelin executes a [REST API call](https://github.com/apache/zeppelin/blob/master/zeppelin-server/src/main/java/org/apache/zeppelin/rest/NotebookRestApi.java) for the notebook permission information.
 In the backend side, Zeppelin gets the user information for the connection and allows the operation if the users and groups
 associated with the current user have at least one entity that belongs to owner entities for the notebook.
+
+## Permissions for folders
+Folder with permissions is way to unite notes with the same permissions and change the permissions together.
+
+Folder without permissions (with empty permissions) continues to work like common folder.
```
Comment (1ambda) [inline docs/setup/security/notebook_authorization.md]: + 1 for this. Will be backward compatible

File: docs/setup/security/notebook_authorization.md (hunk 2 above)
Comment (1ambda) [reply docs/setup/security/notebook_authorization.md]: and What happens if 

- `folder1/note1` has owner
- but `folder1` doesn't have permission
- then anyone can set permission?

File: docs/setup/security/notebook_authorization.md (hunk 3 above)
Comment (tinkoff-dwh) [reply docs/setup/security/notebook_authorization.md]: @1ambda 
what do you mean? if downgrade version?

Comment (1ambda) [reply docs/setup/security/notebook_authorization.md]: I missed `+`, It means I like this behavior.

**Comment by tinkoff-dwh** [comment]
@1ambda 
fixed all items

**Comment by tinkoff-dwh** [comment]
Ready to review

**Comment by tinkoff-dwh** [comment]
Ready to review

**Comment by tinkoff-dwh** [comment]
Ready to review

**Comment by tinkoff-dwh** [comment]
Ready to review

**Comment by tinkoff-dwh** [comment]
need reworking

**Pull request closed without merging**